dapp/src/components/buySell/ApproveSwap.js

```diff
@@ -0,0 +1,257 @@
+import React, { useState, useEffect } from 'react'
+import { fbt } from 'fbt-runtime'
+import { useStoreState } from 'pullstate'
+import { useWeb3React } from '@web3-react/core'
+import ContractStore from 'stores/ContractStore'
+import analytics from 'utils/analytics'
+import withRpcProvider from 'hoc/withRpcProvider'
+import { ethers } from 'ethers'
+import withIsMobile from 'hoc/withIsMobile'
+
+const ApproveSwap = ({
+  stableCoinToApprove,
+  needsApproval,
+  selectedSwap,
+  swapMetadata,
+  onSwap,
+  allowancesLoaded,
+  onMintingError,
+  formHasErrors,
+  swappingGloballyDisabled,
+  storeTransaction,
+  storeTransactionError,
+  rpcProvider,
+  isMobile,
+}) => {
+  const [coinApproved, setCoinApproved] = useState(false)
+  const [stage, setStage] = useState(coinApproved ? 'done' : 'approve')
+  const [contract, setContract] = useState(null)
+  const web3react = useWeb3React()
+  const { library, account } = web3react
+
+  useEffect(() => {
+    const approval = !(
+      (!selectedSwap ||
+        formHasErrors ||
+        swappingGloballyDisabled ||
+        !allowancesLoaded ||
+        !needsApproval) &&
+      !coinApproved
+    )
+    ContractStore.update((s) => {
+      s.approvalNeeded = approval
+    })
+  }, [selectedSwap, needsApproval])
+
+  const {
+    vault,
+    flipper,
+    uniV3SwapRouter,
+    uniV2Router,
+    sushiRouter,
+    curveOUSDMetaPool,
+    usdt,
+    dai,
+    usdc,
+    ousd,
+  } = useStoreState(ContractStore, (s) => s.contracts || {})
+
+  const contracts = {
+    vault: vault,
+    flipper: flipper,
+    uniswap: uniV3SwapRouter,
+    curve: curveOUSDMetaPool,
+    uniswapV2: uniV2Router,
+    sushiswap: sushiRouter,
+  }
+
+  useEffect(() => {
+    setCoinApproved(false)
+    setStage('approve')
+  }, [selectedSwap])
+
+  useEffect(() => {
+    if (stableCoinToApprove === 'dai') {
+      setContract(dai)
+    } else if (stableCoinToApprove === 'usdt') {
+      setContract(usdt)
+    } else if (stableCoinToApprove === 'usdc') {
+      setContract(usdc)
+    } else if (stableCoinToApprove === 'ousd') {
+      setContract(ousd)
+    }
+  }, [stableCoinToApprove, usdt, dai, usdc, ousd])
+
+  const ApprovalMessage = ({ needsApproval, stableCoinToApprove }) => {
+    const capitalized = needsApproval.charAt(0).toUpperCase()
+    const noncapitalized =
+      needsApproval === 'uniswapV2'
+        ? needsApproval.slice(1, 7)
+        : needsApproval.slice(1)
+    const flipper = needsApproval === 'flipper' ? 'the ' : ''
+    const vault = needsApproval === 'vault' ? 'the Origin ' : ''
+    const coin = stableCoinToApprove.toUpperCase()
+    return (
+      <>
+        {isMobile
+          ? fbt(
+              'Approve ' +
+                fbt.param('flipper', flipper) +
+                fbt.param('vault', vault) +
+                fbt.param('capatalized', capitalized) +
+                fbt.param('noncapitalized', noncapitalized),
```

Comment:
It will be easier for translators to translate to another language if there is only 1 `fbt.param`. Since what they see in english is "Approve X" (where X is a variable called by the first parameter in function `fbt.param`). So it makes more sense for them to translate: 
`Approve [coin_contract]`
which in code you would express as
` fbt('Approve ' + fbt.param('coin_contract', '....')`

rather than trying to translate `Approve [flipper][vault][capatalized][noncapitalized]` which current code would produce. 

Can you put the whole coin contract in one variable (and do all the capitalisation, "the" preposition) logic in there. And then insert it as one fbt.param called something that will help transaltors understand what it is. like `coin contract` or `transaction route` for example